File: java-lotto/src/main/java/dev/hongsii/model/WinningResult.java
package dev.hongsii.model;

import java.util.Arrays;

public enum WinningResult {

    FIRST(6, false, 1_000_000),
    SECOND(5, true, 500_000),
    THIRD(5, false, 200_000),
    FOURTH(4, false, 50_000),
    FIFTH(3, false, 5_000),
    NOT_MATCH(0, false, 0);

    private int matchCount;
    private boolean matchBonus;
    private int price;

    WinningResult(int matchCount, boolean matchBonus, int price) {
        this.matchCount = matchCount;
        this.matchBonus = matchBonus;
        this.price = price;
    }

    public static WinningResult of(MatchResult matchResult) {
        return Arrays.stream(values())
                .filter(winningResult -> winningResult.parse(matchResult))
                .findFirst()
                .orElse(NOT_MATCH);
    }

    private boolean parse(MatchResult matchResult) {
        boolean isSameCount = (matchCount == matchResult.getMatchCount());
        if (this == SECOND) {
            return isSameCount && (matchBonus == matchResult.matchBonus());
        }
        return isSameCount;
    }

    public int getMatchCount() {
        return matchCount;
    }

    public boolean isMatchBonus() {
        return matchBonus;
    }

    public int getPrice() {
        return price;
    }
}

File: kotlin-lotto/src/main/kotlin/lotto/model/WinningResult.kt
package lotto.model

enum class WinningResult(private val matchResult: MatchResult, val winningMoney: Int) {

    FIRST(6, 1_000_000),
    SECOND(5, true, 500_000),
    THIRD(5, 200_000),
    FOURTH(4, 50_000),
    FIFTH(3, 5_000),
    NONE(0, 0);

    constructor(matchCount: Int, winningMoney: Int) : this(matchCount, false, winningMoney)
    constructor(matchCount: Int, matchBonusNumber: Boolean, winningMoney: Int) : this(MatchResult(matchCount, matchBonusNumber), winningMoney)

    companion object {
        fun of(matchResult: MatchResult) = values().firstOrNull { it.isSameResult(matchResult) } ?: NONE
    }

    fun isSameResult(matchResult: MatchResult): Boolean = if (this.matchResult.matchBonusNumber) {
        this.matchResult == matchResult
    } else {
        this.matchResult.isSameMatchCount(matchResult)
    }
}
